File: cli/lib/user/crypto.js
'use strict';

var user = exports;

var kdf = require('../crypto/kdf');
var utils = require('../crypto/utils');
var triplesec = require('../crypto/triplesec');

var algos = require('common/types/algos');
var base64url = require('base64url');

var SALT_BYTES = 16;
var MASTER_KEY_BYTES = 256;
var SLICE_LENGTH_BYTES = 192;
var REMAINDER_SLICE_LENGTH_BYTES = 32;

/**
 * Generate both the password and master objects for user.body
 *
 * A 128bit salt is generated / used to encrypt the plaintext password
 * The encrypted password buffer is sliced [192] and encoded in base64
 * A 1024bit key is generated as used as the master key
 * The encrypted password buffer is sliced [0,192] / used to encrypt master key
 * The encrypted+encoded values are stored on the server
 *
 * @param {string} password - Plaintext password value
 */
user.encryptPasswordObject = function(password) {
  var data = {};

  // Generate 128 bit (16 byte) salt for password
  return utils.randomBytes(SALT_BYTES)
    // Construct the password object
    .then(function(passwordSalt) {
      data.password = {
        salt: base64url.encode(passwordSalt),
        alg: algos.value('scrypt'), // 0x23
      };

      // Create password buffer
      return kdf.generate(password, passwordSalt).then(function(buf) {
        // Append the base64url value
        data.password.value = user.pwh(buf);
        return buf;
      });

    // Construct the master object from the password buffer
    }).then(function(passwordBuf) {
      data.master = {
        alg: algos.value('triplesec v3'), // 0x22
      };

      // Generate 1024 bit (256 byte) master key
      return utils.randomBytes(MASTER_KEY_BYTES).then(function(masterKeyBuf) {
        var passwordBufSlice = passwordBuf.slice(0, SLICE_LENGTH_BYTES);
        // Ensure the buffer used to encrypt is 192 bytes
        if (passwordBufSlice.length !== SLICE_LENGTH_BYTES) {
          throw new Error('invalid buffer length');
        }
        // Encrypt master key using the password buffer
        return triplesec.encrypt({
          data: masterKeyBuf,
          key: passwordBufSlice,
        }).then(function(buf) {
          // Base64 the master value for transmission
          data.master.value = base64url.encode(buf);
          return data;
        });
      });
    });
};

/**
 * Slice password and encode
 *
 * @param {buffer} passwordBuf
 */
user.pwh = function(passwordBuf) {
  // pwh is created from last 32 byte / 256 bit of password buffer
  var passwordBufSlice = passwordBuf.slice(SLICE_LENGTH_BYTES);
  if (passwordBufSlice.length !== REMAINDER_SLICE_LENGTH_BYTES) {
    throw new Error('invalid buffer length');
  }
  return base64url.encode(passwordBufSlice);
};

/**
 * Decrypt the master key from user object using password
 *
 * @param {string} password - Plaintext password
 * @param {object} userObject - Full user object
 */
user.decryptMasterKey = function(password, userObject) {
  // Use stored password salt to encrypt password
  var passwordSalt = userObject.body.password.salt;
  return kdf.generate(password, passwordSalt).then(function(buf) {
    // Decode master value from base64url
    var value = base64url.toBuffer(userObject.body.master.value);
    // Use the password buffer to decrypt the master key
    // 192 byte slice
    var masterKey = buf.slice(0, SLICE_LENGTH_BYTES);
    if (masterKey.length !== SLICE_LENGTH_BYTES) {
      throw new Error('invalid buffer length');
    }
    // Returns masterKey buffer for use with encrypting
    return triplesec.decrypt({
      data: value,
      key: masterKey,
    });
  });
};

File: daemon/crypto/crypto.go
package crypto

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha512"
	"encoding/base64"

	"golang.org/x/crypto/ed25519"
	"golang.org/x/crypto/scrypt"

	base64url "github.com/manifoldco/torus-cli/base64"
	"github.com/manifoldco/torus-cli/daemon/ctxutil"
	"github.com/manifoldco/torus-cli/primitive"
)

// Crypto Algorithm name constants.
const (
	Triplesec  = "triplesec-v3"
	EdDSA      = "eddsa"
	Curve25519 = "curve25519"
	EasyBox    = "easybox"
	SecretBox  = "secretbox"
	Scrypt     = "scrypt"
)

// scrypt parameter constants
const (
	n              = 32768 // 2^15
	r              = 8
	p              = 1
	keyLen         = 256
	saltBytes      = 16
	masterKeyBytes = 256
)

// LoginKeypair represents an Ed25519 Keypair used for generating a login token
// signature for Passphrase-Dervied Public Key Authentication.
type LoginKeypair struct {
	public  ed25519.PublicKey
	private ed25519.PrivateKey
	salt    *base64url.Value
}

// PublicKey returns the base64 value of the public key
func (k *LoginKeypair) PublicKey() *base64url.Value {
	return base64url.NewValue(k.public)
}

// Salt returns the base64 representation of the salt used to derive the
// LoginKeypair
func (k *LoginKeypair) Salt() *base64url.Value {
	return k.salt
}

// Sign returns a signature of the given token as a base64 string
func (k *LoginKeypair) Sign(token []byte) *base64url.Value {
	sig := ed25519.Sign(k.private, token)
	return base64url.NewValue(sig)
}

// DeriveLoginHMAC HMACs the provided token with a key derived from password
// and the provided base64 encoded salt.
func DeriveLoginHMAC(ctx context.Context, password []byte, salt, token string) (string, error) {
	key, err := derivePassword(ctx, password, salt)
	if err != nil {
		return "", err
	}

	pwh := make([]byte, base64.RawURLEncoding.EncodedLen(32))
	base64.RawURLEncoding.Encode(pwh, key)

	mac := hmac.New(sha512.New, pwh)
	mac.Write([]byte(token))
	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil)), nil
}

func deriveHash(ctx context.Context, password []byte, salt string) ([]byte, error) {
	s := make([]byte, base64.RawURLEncoding.DecodedLen(len(salt)))
	l, err := base64.RawURLEncoding.Decode(s, []byte(salt))
	if err != nil {
		return nil, err
	}

	err = ctxutil.ErrIfDone(ctx)
	if err != nil {
		return nil, err
	}

	key, err := scrypt.Key(password, s[:l], n, r, p, keyLen)
	return key, err
}

func derivePassword(ctx context.Context, password []byte, salt string) ([]byte, error) {
	key, err := deriveHash(ctx, password, salt)
	if err != nil {
		return nil, err
	}

	pwh := key[192:224]
	err = ctxutil.ErrIfDone(ctx)
	if err != nil {
		return nil, err
	}

	return pwh, err
}

// GenerateSalt returns a 16-byte (128 bit) salt used in password and secret
// key derivation.
func GenerateSalt(ctx context.Context) (*base64url.Value, error) {
	err := ctxutil.ErrIfDone(ctx)
	if err != nil {
		return nil, err
	}

	salt := make([]byte, saltBytes) // 16
	_, err = rand.Read(salt)
	if err != nil {
		return nil, err
	}

	return base64url.NewValue(salt), nil
}

// EncryptPasswordObject derives the master key and password hash from password
// and salt, returning the master and password objects
func EncryptPasswordObject(ctx context.Context, password string) (*primitive.UserPassword, *primitive.MasterKey, error) {
	pw := &primitive.UserPassword{
		Alg: Scrypt,
	}

	// Generate 128 bit (16 byte) salt for password
	salt := make([]byte, saltBytes) // 16
	_, err := rand.Read(salt)
	if err != nil {
		return nil, nil, err
	}

	// Encode salt bytes to base64url
	pw.Salt = base64.RawURLEncoding.EncodeToString(salt)

	// Create password hash bytes
	pwh, err := derivePassword(ctx, []byte(password), pw.Salt)
	if err != nil {
		return nil, nil, err
	}

	// Encode password value to base64url
	pw.Value = base64url.NewValue(pwh)

	m, err := CreateMasterKeyObject(ctx, []byte(password))
	if err != nil {
		return nil, nil, err
	}

	return pw, m, nil
}

// CreateMasterKeyObject generates a 256 byte master key which is then
// encrypted using TripleSec-v3 using the given password.
func CreateMasterKeyObject(ctx context.Context, password []byte) (*primitive.MasterKey, error) {
	m := &primitive.MasterKey{
		Alg: Triplesec,
	}

	// Generate a master key of 256 bytes
	key := make([]byte, masterKeyBytes)
	_, err := rand.Read(key)
	if err != nil {
		return nil, err
	}

	err = ctxutil.ErrIfDone(ctx)
	if err != nil {
		return nil, err
	}

	ts, err := newTriplesec(ctx, password)
	if err != nil {
		return nil, err
	}

	ct, err := ts.Encrypt(key)
	if err != nil {
		return nil, err
	}

	// Encode the master key value to base64url
	m.Value = base64url.NewValue(ct)
	return m, nil
}

// DeriveLoginKeypair dervies the ed25519 login keypair used for machine
// authentication from the given salt and secret values.
func DeriveLoginKeypair(ctx context.Context, secret, salt *base64url.Value) (
	*LoginKeypair, error) {

	key, err := deriveHash(ctx, *secret, salt.String())
	if err != nil {
		return nil, err
	}

	err = ctxutil.ErrIfDone(ctx)
	if err != nil {
		return nil, err
	}

	r := bytes.NewReader(key[224:]) // Use last 32 bytes of 256 to derive key
	pubKey, privKey, err := ed25519.GenerateKey(r)
	if err != nil {
		return nil, err
	}

	err = ctxutil.ErrIfDone(ctx)
	if err != nil {
		return nil, err
	}

	keypair := &LoginKeypair{
		public:  pubKey,
		private: privKey,
		salt:    salt,
	}
	return keypair, nil
}
